src/coreclr/jit/morph.cpp

```diff
@@ -2835,6 +2846,33 @@ void CallArgs::AddFinalArgsAndDetermineABIInfo(Compiler* comp, GenTreeCall* call
                         arg.AbiInfo.StructFloatFieldType[1] =
                             (floatFieldFlags & STRUCT_SECOND_FIELD_SIZE_IS8) ? TYP_DOUBLE : TYP_FLOAT;
                     }
+
+                    if (isGCTypes)
```

Comment:
With the `!varTypeIsGC(getSlotType(inx))` check in multi-reg morphing, there is actually no need to set the GC types here, except for the code in `genCall`, which can query the `PUTARG_SPLIT` node instead for the right register type.

The way this is currently written, the contract for when `StructFloatFieldType` gets set is unnecessarily complex, e. g. it will be `{ TYP_REF, TYP_UNDEF }` for a split `struct { object, long }`, but not for a register-passed struct of same kind.

We should either delete the handling of GC types here, or make it "complete". I would personally have a preference towards the former, since its less code.

Reply:
I can‘t express my viewpoint very well in english.

> With the `!varTypeIsGC(getSlotType(inx))` check in multi-reg morphing, there is actually no need to set the GC types here, except for the code in `genCall`, which can query the `PUTARG_SPLIT` node instead for the right register type.

Do you mean getting the GC types within the `genCall` for the split case ?

Reply:
> Do you mean getting the GC types within the `genCall` for the split case ?

Yes.

Reply:
By the way, I have another question about the second field of the split which passed on stack within the `fgMorphMultiregStructArg` ?

Reply:
> By the way, I have another question about the second field of the split which passed on stack within the `fgMorphMultiregStructArg` ?

Yes, what would it be about?

I believe the code for that case was and is correct.

Reply:
> > By the way, I have another question about the second field of the split which passed on stack within the `fgMorphMultiregStructArg` ?
> 
> Yes, what would it be about?

How to process the second field as the elemCount is 2 ?

Reply:
Is there a problem with what the code currently does? I believe it should be correct; the split case is treated the same as a two-reg case by multi-reg morphing.

Reply:
> Is there a problem with what the code currently does? 

No.

> I believe it should be correct; 

Yes.

> the split case is treated the same as a two-reg one by multi-reg morphing.

Yes, as a two-reg but will ignor the second ?

Reply:
> Yes, as a two-reg but will ignore the second?

Why would it? It should always create a `FIELD_LIST` with two items.

Reply:
ooh, I see. Thanks.

> It should always create a `FIELD_LIST` with two items.

yes, it will be treated as stack slot.

Reply:
@SingleAccretion   Thanks very much again for your help, I gain a lot~